File: leetcode/easy/1768.merge-strings-alternately/solutions/go/main.go
/**
 * Main executable for Merge Strings Alternately problem
 *
 * This file contains the main function with sample usage and comprehensive testing.
 */
package main

import (
	"fmt"
	"strings"
	"time"
)

// Main function with sample usage and test cases
func main() {
	fmt.Println("=== Merge Strings Alternately - Go Solution ===")

	// Test cases from problem description
	testCases := []struct {
		word1, word2, expected string
		description            string
	}{
		{"abc", "pqr", "apbqcr", "Example 1: Equal length strings"},
		{"ab", "pqrs", "apbqrs", "Example 2: word2 longer"},
		{"abcd", "pq", "apbqcd", "Example 3: word1 longer"},
		{"a", "b", "ab", "Single characters"},
		{"", "abc", "abc", "Empty word1"},
		{"abc", "", "abc", "Empty word2"},
		{"", "", "", "Both empty"},
		{"aaa", "bbb", "ababab", "Repeated characters"},
		{"aceg", "bdfh", "abcdefgh", "Alternating pattern"},
	}

	// Run test cases
	fmt.Println("--- Test Cases ---")
	for _, tc := range testCases {
		result := mergeAlternately(tc.word1, tc.word2)
		status := "✅ PASS"
		if result != tc.expected {
			status = "❌ FAIL"
		}

		fmt.Printf("Test: %s\n", tc.description)
		fmt.Printf("Input: word1=\"%s\", word2=\"%s\"\n", tc.word1, tc.word2)
		fmt.Printf("Output: \"%s\"\n", result)
		fmt.Printf("Expected: \"%s\"\n", tc.expected)
		fmt.Printf("Result: %s\n\n", status)
	}

	// Performance test
	performanceTest()

	// Edge cases test
	edgeCasesTest()

	// Memory usage test
	memoryUsageTest()
}

// performanceTest measures execution time with large inputs
func performanceTest() {
	fmt.Println("--- Performance Test ---")

	// Create large test data
	word1 := strings.Repeat("a", 1000)
	word2 := strings.Repeat("b", 1000)

	// Warm up
	for i := 0; i < 1000; i++ {
		mergeAlternately("a", "b")
	}

	// Measure performance
	iterations := 10000
	startTime := time.Now()

	for i := 0; i < iterations; i++ {
		mergeAlternately(word1, word2)
	}

	endTime := time.Now()
	duration := endTime.Sub(startTime)

	fmt.Printf("Test data: word1=\"%s\" (%d chars), word2=\"%s\" (%d chars)\n",
		word1[:min(10, len(word1))]+"...", len(word1),
		word2[:min(10, len(word2))]+"...", len(word2))
	fmt.Printf("Executed %d iterations in %v\n", iterations, duration)
	fmt.Printf("Average time per iteration: %v\n", duration/time.Duration(iterations))
	fmt.Printf("Operations per second: %.0f\n\n", float64(iterations)/duration.Seconds())
}

// edgeCasesTest tests boundary conditions and edge cases
func edgeCasesTest() {
	fmt.Println("--- Edge Cases Test ---")

	edgeCases := []struct {
		word1, word2, expected string
		description            string
	}{
		{strings.Repeat("a", 100), strings.Repeat("b", 100), strings.Repeat("ab", 100), "Maximum length strings (100 chars each)"},
		{"a", strings.Repeat("b", 99), "a" + strings.Repeat("b", 99), "Very different lengths (1 vs 99)"},
		{strings.Repeat("a", 50), strings.Repeat("b", 100), strings.Repeat("ab", 50) + strings.Repeat("b", 50), "Unequal maximum lengths"},
	}

	for _, tc := range edgeCases {
		result := mergeAlternately(tc.word1, tc.word2)
		status := "✅ PASS"
		if result != tc.expected {
			status = "❌ FAIL"
		}

		fmt.Printf("Test: %s\n", tc.description)
		fmt.Printf("Input lengths: %d + %d = %d\n", len(tc.word1), len(tc.word2), len(tc.word1)+len(tc.word2))
		fmt.Printf("Output length: %d\n", len(result))
		fmt.Printf("Result: %s\n\n", status)
	}

	fmt.Println("All edge cases completed successfully! ✅")
}

// memoryUsageTest demonstrates memory efficiency
func memoryUsageTest() {
	fmt.Println("--- Memory Usage Test ---")

	// Test with large inputs
	word1 := strings.Repeat("a", 1000)
	word2 := strings.Repeat("b", 1000)

	fmt.Printf("Input: word1=\"%s\" (%d chars), word2=\"%s\" (%d chars)\n",
		word1[:min(10, len(word1))]+"...", len(word1),
		word2[:min(10, len(word2))]+"...", len(word2))

	result := mergeAlternately(word1, word2)

	fmt.Printf("Output: \"%s\" (%d chars)\n",
		result[:min(20, len(result))]+"...", len(result))
	fmt.Printf("Memory efficiency: Result length = %d (expected: %d)\n",
		len(result), len(word1)+len(word2))

	if len(result) == len(word1)+len(word2) {
		fmt.Println("✅ Memory usage is optimal")
	} else {
		fmt.Println("❌ Memory usage issue detected")
	}

	fmt.Println()
}

// min returns the minimum of two integers
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: leetcode/easy/1768.merge-strings-alternately/solutions/java/Main.java
/**
 * Main executable class for Merge Strings Alternately problem
 * 
 * This class demonstrates both solutions with comprehensive test cases
 * and performance measurements.
 */
public class Main {
    public static void main(String[] args) {
        System.out.println("=== Merge Strings Alternately - Java Solutions ===\n");
        
        // Test both solutions
        testSolution1();
        testSolution2();
        
        // Performance comparison
        performanceComparison();
        
        // Edge cases testing
        testEdgeCases();
    }
    
    /**
     * Test Solution 1: StringBuilder with Two Pointers
     */
    private static void testSolution1() {
        System.out.println("--- Solution 1: StringBuilder with Two Pointers ---");
        Solution1 solution = new Solution1();
        
        // Test cases from problem description
        testCase(solution, "abc", "pqr", "apbqcr", "Example 1: Equal length strings");
        testCase(solution, "ab", "pqrs", "apbqrs", "Example 2: word2 longer");
        testCase(solution, "abcd", "pq", "apbqcd", "Example 3: word1 longer");
        
        // Additional test cases
        testCase(solution, "a", "b", "ab", "Single characters");
        testCase(solution, "", "abc", "abc", "Empty word1");
        testCase(solution, "abc", "", "abc", "Empty word2");
        testCase(solution, "", "", "", "Both empty");
        
        System.out.println();
    }
    
    /**
     * Test Solution 2: Character Array with Bulk Copy
     */
    private static void testSolution2() {
        System.out.println("--- Solution 2: Character Array with Bulk Copy ---");
        Solution2 solution = new Solution2();
        
        // Test cases from problem description
        testCase(solution, "abc", "pqr", "apbqcr", "Example 1: Equal length strings");
        testCase(solution, "ab", "pqrs", "apbqrs", "Example 2: word2 longer");
        testCase(solution, "abcd", "pq", "apbqcd", "Example 3: word1 longer");
        
        // Additional test cases
        testCase(solution, "a", "b", "ab", "Single characters");
        testCase(solution, "", "abc", "abc", "Empty word1");
        testCase(solution, "abc", "", "abc", "Empty word2");
        testCase(solution, "", "", "", "Both empty");
        
        System.out.println();
    }
    
    /**
     * Test a single case for any solution
     */
    private static void testCase(Object solution, String word1, String word2, String expected, String description) {
        String result;
        
        // Call the appropriate solution method
        if (solution instanceof Solution1) {
            result = ((Solution1) solution).mergeAlternately(word1, word2);
        } else if (solution instanceof Solution2) {
            result = ((Solution2) solution).mergeAlternately(word1, word2);
        } else {
            throw new IllegalArgumentException("Unknown solution type");
        }
        
        // Display results
        System.out.printf("Test: %s%n", description);
        System.out.printf("Input: word1=\"%s\", word2=\"%s\"%n", word1, word2);
        System.out.printf("Output: \"%s\"%n", result);
        System.out.printf("Expected: \"%s\"%n", expected);
        System.out.printf("Result: %s%n%n", result.equals(expected) ? "✅ PASS" : "❌ FAIL");
    }
    
    /**
     * Performance comparison between both solutions
     */
    private static void performanceComparison() {
        System.out.println("--- Performance Comparison ---");
        
        // Create test data
        String word1 = "a".repeat(1000);
        String word2 = "b".repeat(1000);
        
        Solution1 solution1 = new Solution1();
        Solution2 solution2 = new Solution2();
        
        // Warm up JVM
        for (int i = 0; i < 1000; i++) {
            solution1.mergeAlternately("a", "b");
            solution2.mergeAlternately("a", "b");
        }
        
        // Test Solution 1
        long startTime = System.nanoTime();
        for (int i = 0; i < 10000; i++) {
            solution1.mergeAlternately(word1, word2);
        }
        long endTime = System.nanoTime();
        double solution1Time = (endTime - startTime) / 1_000_000.0; // Convert to milliseconds
        
        // Test Solution 2
        startTime = System.nanoTime();
        for (int i = 0; i < 10000; i++) {
            solution2.mergeAlternately(word1, word2);
        }
        endTime = System.nanoTime();
        double solution2Time = (endTime - startTime) / 1_000_000.0; // Convert to milliseconds
        
        System.out.printf("Test data: word1=\"%s\" (%d chars), word2=\"%s\" (%d chars)%n", 
            word1.substring(0, Math.min(10, word1.length())) + "...", word1.length(),
            word2.substring(0, Math.min(10, word2.length())) + "...", word2.length());
        System.out.printf("Solution 1 (StringBuilder): %.2f ms for 10,000 iterations%n", solution1Time);
        System.out.printf("Solution 2 (Character Array): %.2f ms for 10,000 iterations%n", solution2Time);
        System.out.printf("Performance ratio: %.2fx%n%n", solution1Time / solution2Time);
    }
    
    /**
     * Test edge cases and boundary conditions
     */
    private static void testEdgeCases() {
        System.out.println("--- Edge Cases Testing ---");
        
        Solution1 solution = new Solution1();
        
        // Test with maximum constraint values
        String maxWord1 = "a".repeat(100);
        String maxWord2 = "b".repeat(100);
        String expectedMax = "ab".repeat(100);
        
        testCase(solution, maxWord1, maxWord2, expectedMax, "Maximum length strings (100 chars each)");
        
        // Test with very different lengths
        String shortWord = "a";
        String longWord = "b".repeat(99);
        String expectedDifferent = "a" + "b".repeat(99);
        
        testCase(solution, shortWord, longWord, expectedDifferent, "Very different lengths (1 vs 99)");
        
        // Test with repeated characters
        String repeated1 = "aaa";
        String repeated2 = "bbb";
        String expectedRepeated = "ababab";
        
        testCase(solution, repeated1, repeated2, expectedRepeated, "Repeated characters");
        
        // Test with alternating pattern
        String pattern1 = "aceg";
        String pattern2 = "bdfh";
        String expectedPattern = "abcdefgh";
        
        testCase(solution, pattern1, pattern2, expectedPattern, "Alternating pattern");
        
        System.out.println("All edge cases completed successfully! ✅");
    }
    
    /**
     * Memory usage measurement
     */
    private static void measureMemoryUsage() {
        System.out.println("--- Memory Usage Analysis ---");
        
        Runtime runtime = Runtime.getRuntime();
        
        // Force garbage collection
        System.gc();
        long beforeMemory = runtime.totalMemory() - runtime.freeMemory();
        
        // Execute solution with large input
        Solution1 solution = new Solution1();
        String result = solution.mergeAlternately("a".repeat(1000), "b".repeat(1000));
        
        long afterMemory = runtime.totalMemory() - runtime.freeMemory();
        long memoryUsed = afterMemory - beforeMemory;
        
        System.out.printf("Memory used for 1000+1000 character strings: %d bytes (%.2f KB)%n", 
            memoryUsed, memoryUsed / 1024.0);
        System.out.printf("Result length: %d characters%n", result.length());
        System.out.println();
    }
}
